### ISSUE TITLE
[Improvement] Improve accessibility for screen reader users

### ISSUE THREAD

**Issue by Neurrone** [Kong/insomnia#1436]
- Insomnia Version: latest, 6.4.3
- Operating System: Windows 10 1809__

Insomnia works fairly well out of the box, since most elements read correctly, but there are some things that could be improved.

All testing done with the [NVDA screen reader](https://www.nvaccess.org/).

#### Offscreen content is still visible to screen readers

There is some content that is visually hidden that still shows up for me with NVDA. After the preview pane that shows the received response, I'm seeing the following:

1. A `<h2>loading</h2>`, followed by a "cancel request button"
2. Some sort of find dialog that shows "No match :(" if no matches are found. This field also keeps stealing keyboard focus periodically, making it harder to use the app.
3. Some extra information about recently received responses, e.g "The request has succeeded. 1203.000 milliseconds Read 73.4 KB Content"

#### Contents of editable fields can't be read

Nothing is announced when using the arrows while moving about through editable fields such as when entering url, or when reading the received response, which looks like a problem caused by [how CodeMirror uses the DOM](https://github.com/codemirror/CodeMirror/issues/4604)

**Comment by gschier** [comment]
Thanks for taking the time to write this up! I will add `aria-hidden` attributes to the hidden content right now since that is straightforward. 

For the editable fields, we might be blocked on this CodeMirror issue: https://github.com/codemirror/CodeMirror/issues/4604

**Comment by Neurrone** [comment]
Since this offscreen content is meant to b hidden for everyone, i suggest using the hidden attribute instead.

There is a recemt firefox patch that fixes the code mirror issue. Will look for it

Sent from my iPhone

________________________________

**Comment by gschier** [comment]
Ah, thanks! I didn't know about the `hidden` attribute.

**Comment by Neurrone** [comment]
@gschier thanks for being willing to help with this. 

forgot to also mention that `display: none` would have worked too. See this [article on hiding content](https://www.scottohara.me/blog/2017/04/14/inclusively-hidden.html) for more info.

Pending CodeMirror 6 which should fix the editor, there is a workaround implemented in Firefox now to allow codemirror editors to work - this is [the MR introducing the change](https://phabricator.services.mozilla.com/D24930) and [the associated tracking bug](https://bugzilla.mozilla.org/show_bug.cgi?id=919711). This works by loading code that adds a new input style, and using it when configuring CodeMirror.

**Comment by zersiax** [comment]
I just had a play with the latest Insomnia as well. While I haven't quite found the issues @Neurrone  describes, I did find a number of controls that aren't labeled or could use some ARIA love in the request creation UI:

- The HTTP VERB selector doesn't currently indicate that it can be expanded. It just lists as a button with the label "GET" which can be anything really. Adding aria-expanded  to this and toggling it with JS when it toggles would already fix this.
- Various buttons in, for example, the headers tab, as well as various edit fields currently have no label that screenreaders can immediately discover. If there are labels in the UI, aria-labeledby="your_element"here"  would already fix this issue. If not, consider adding labels, or add aria-label="blah" to controls that currently only have an icon.

Insomnia is already much, much easier to use than some alternatives, and these fixes really are super low-hanging fruit. Me and many others would really appreciate some of these things getting fixed.
As for the CodeMirror input fields, the last I've heard on that is that the new version would address accessibility concerns, but it has been a while since I've heard about any progress on that front. I'll go check on that after I'm done here :-) For an anternative, Monaco might be an option which is significantly more accessible, but I realize that isn't exactly a small ask :)

**Comment by zersiax** [comment]
Plaintive bump? :) 
@gschier I added some feedback on this issue two weeks ago, just making sure you saw it? :)

**Comment by ajtran303** [comment]
@zersiax @gschier 

Hi y'all! I would be willing to work on this.

**Comment by zersiax** [comment]
@ajtran303  good to hear, do you need anything info or testing-wise to be able to? :)

**Comment by ajtran303** [comment]
> @ajtran303 good to hear, do you need anything info or testing-wise to be able to? :)

@zersiax I will keep you updated. I just made a fork of this repo and will learn where the relevant code lives.

I have followed the discussion about adding ARIA labels to the UI.

But I don't quite understand about the previous comments about CodeMirror.

**Comment by zersiax** [comment]
@ajtran303   Versions of CodeMirror that aren't the most recent v6 suffer from an accessibility issue that makes screenreaders unable to properly read the contents of the editor, which makes working with those editors to look at responses and fill in request params extremely annoying. I have no idea if CodeMirror is still even being used , but if so, that will likely mean the edit fields are still not properly readable. The only way is to copy-paste the contents out of the field and into a notepad window or similar, which is a bit of a productivity hit.

**Comment by ajtran303** [comment]
Hey @zersiax so there has been a UI update since I have last used this client. #3183 #3178 

I think I have figured out how to address the issue of adding `aria-expanded` attributes to the drop-down components but overall I feel discouraged now since there are even more navigation problems with the app.

I've been searching for an accessible HTTP client that I can comfortably recommend to people to download and use and I'm not sure if I will recommend Insomnia anymore. In the meantime, I can write down instructions for how to try out my branch and navigate to the parts of the app that have drop-downs -- can you test to check if the `aria-expanded` attribute is helping to indicate that there is a drop-down?
